2.0/m_joinpartspam.cpp

```diff
@@ -0,0 +1,231 @@
+/*
+ * InspIRCd -- Internet Relay Chat Daemon
+ *
+ *   Copyright (C) 2016 genius3000 <genius3000@g3k.solutions>
+ *
+ * This file is a module for InspIRCd.  InspIRCd is free software: you can
+ * redistribute it and/or modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation, version 2.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "inspircd.h"
+
+/* $ModDesc: Adds channel mode +V to ban a user after x per y joins and parts/quits (join/part spam) */
+/* $ModAuthor: genius3000 */
+/* $ModAuthorMail: genius3000@g3k.solutions */
+/* $ModDepends: core 2.0 */
+
+class joinpartspamsettings
+{
+ public:
+	unsigned int cycles;
+	unsigned int secs;
+	time_t lastcleanup;
+
+	struct Tracking
+	{
+		time_t reset;
+		unsigned int counter;
+	};
+	std::map<std::string, Tracking> cycler;
+
+	joinpartspamsettings(int c, int s) : cycles(c), secs(s)
+	{
+	}
+
+	/* Called by PostJoin to possibly reset a cycler's Tracking and increment the counter */
+	void addcycle(const std::string& mask)
+	{
+		/* If mask isn't already tracked, set reset time
+		 * If tracked and reset time is up, reset counter and reset time
+		 * Also assume another server denied join and set ban, with a user removing
+		 * the ban if counter >= cycles, reset counter and reset time
+		 */
+		if (cycler.count(mask) != 1)
+			cycler[mask].reset = ServerInstance->Time() + secs;
+		else if (ServerInstance->Time() > cycler[mask].reset || cycler[mask].counter >= cycles)
+		{
+			cycler[mask].counter = 0;
+			cycler[mask].reset = ServerInstance->Time() + secs;
+		}
+
+		++cycler[mask].counter;
```

Comment:
Each `cycler[mask]` causes a map lookup which is wasteful because you are only interacting with the same element throughout the method. To improve this, do something like:
`Tracking& tracking = cycler[mask];`
then replace `cycler[mask]` with `tracking` in the rest of the method.